runtime/browser/xwalk_notification_manager_linux.cc

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2015 Intel Corporation. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "xwalk/runtime/browser/xwalk_notification_manager_linux.h"
+
+#include <map>
+
+#include "base/strings/utf_string_conversions.h"
+#include "content/public/browser/browser_thread.h"
+#include "content/public/browser/desktop_notification_delegate.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/common/platform_notification_data.h"
+#include "third_party/skia/include/core/SkBitmap.h"
+#include "url/gurl.h"
+
+namespace {
+
+using xwalk::XWalkNotificationManagerLinux;
+using content::BrowserThread;
+
+// definded by 'org.freedesktop.Notifications'
+#define ND_NOTIFICATION_CLOSED_USER 2
+
+const char gAppNameForNotification[] = "xwalk";
+
+void NotificationClosedCallback(NotifyNotification* notification,
+                                gpointer user_data) {
+  XWalkNotificationManagerLinux* service =
+    static_cast<XWalkNotificationManagerLinux*>(user_data);
+
+  gint reason = notify_notification_get_closed_reason(notification);
+
+  bool by_user = false;
+  if (reason == ND_NOTIFICATION_CLOSED_USER) {
+    service->NotificationClicked(notification);
+    by_user = true;
+  }
+  service->NotificationClosed(notification, by_user);
+  g_signal_handler_disconnect(notification,
+                              service->closed_handler(notification));
+}
+
+void CancelDesktopNotification(
+        XWalkNotificationManagerLinux* service,
+        NotifyNotification* notification) {
+  g_signal_handler_disconnect(notification,
+                              service->closed_handler(notification));
+  notify_notification_close(notification, nullptr);
+}
+
+}  // namespace
+
+namespace xwalk {
+
+XWalkNotificationManagerLinux::XWalkNotificationManagerLinux() :
+                                initted_(false) {
+  if (notify_init(gAppNameForNotification))
+    initted_ = true;
+}
+
+XWalkNotificationManagerLinux::~XWalkNotificationManagerLinux() {
+  if (initted_)
+    notify_uninit();
+  notifications_map_.clear();
+  notifications_replace_map_.clear();
+  notifications_handler_map_.clear();
+}
+
+void XWalkNotificationManagerLinux::ShowDesktopNotification(
+        content::BrowserContext* browser_context,
+        const GURL& origin,
+        const content::PlatformNotificationData& notification_data,
+        scoped_ptr<content::DesktopNotificationDelegate> delegate,
+        int render_process_id,
+        base::Closure* cancel_callback) {
+  if (!initted_)
+      return;
+
+  NotifyNotification* notification = nullptr;
+
+  if (!notification_data.tag.empty() &&
+        notifications_replace_map_[notification_data.tag]) {
```

Comment:
don't use [] operator without making sure the key value pair exists within the container